### ISSUE TITLE
Use of auto_ptr causes dirty compile under Cygwin with -std=c++11

### ISSUE THREAD

**Issue by noloader** [weidai11/cryptopp#17]
cryptlib.cpp: In member function ‘virtual size_t CryptoPP::PK_Signer::SignMessage(CryptoPP::RandomNumberGenerator&, const byte*, size_t, byte*) const’:
    cryptlib.cpp:770:41: warning: ‘auto_ptr’ is deprecated (declared at /usr/lib/gcc/x86_64-pc-cygwin/4.9.3/include/c++/backward/auto_ptr.h:87) [-Wdeprecated-declarations]
      std::auto_ptr<PK_MessageAccumulator> m(NewSignatureAccumulator(rng));
                                             ^
    cryptlib.cpp: In member function ‘virtual size_t     CryptoPP::PK_Signer::SignMessageWithRecovery(CryptoPP::RandomNumberGenerator&, const byte*, size_t, const byte*, size_t, byte*) const’:
    cryptlib.cpp:778:41: warning: ‘auto_ptr’ is deprecated (declared at /usr/lib/gcc/x86_64-pc-cygwin/4.9.3/include/c++/backward/auto_ptr.h:87) [-Wdeprecated-declarations]
      std::auto_ptr<PK_MessageAccumulator> m(NewSignatureAccumulator(rng));
                                             ^
    cryptlib.cpp: In member function ‘virtual bool     CryptoPP::PK_Verifier::Verify(CryptoPP::PK_MessageAccumulator*) const’:
    cryptlib.cpp:786:59: warning: ‘auto_ptr’ is deprecated (declared at /usr/lib/gcc/x86_64-pc-cygwin/4.9.3/include/c++/backward/auto_ptr.h:87) [-Wdeprecated-declarations]
      std::auto_ptr<PK_MessageAccumulator> m(messageAccumulator);
                                                               ^
    cryptlib.cpp: In member function ‘virtual bool CryptoPP::PK_Verifier::VerifyMessage(const byte*, size_t, const byte*, size_t) const’:
    cryptlib.cpp:792:41: warning: ‘auto_ptr’ is deprecated (declared at /usr/lib/gcc/x86_64-pc-cygwin/4.9.3/include/c++/backward/auto_ptr.h:87) [-Wdeprecated-declarations]
    std::auto_ptr<PK_MessageAccumulator> m(NewVerificationAccumulator());
                                             ^
    ...

**Comment by denisbider** [comment]
It seems challenging to simultaneously compile with no warnings both on C++11 platforms that deprecate auto_ptr, and C++03 platforms that don't have unique_ptr. Is there a way to disable the warning?

**Comment by noloader** [comment]
Cleared through a few different commits. Some are listed below.

The general strategy was to identify C++11, identify ancient C++11 standard libraries, and then swap-in `unique_ptr` for `auto_ptr` if available. The identification and swap-in occurred in `smartptr.h`:

    // Hack ahead. Apple's standard library does not have C++'s unique_ptr in C++11. We can't
    //   test for unique_ptr directly because some of the non-Apple Clangs on OS X fail the same
    //   way. However, modern standard libraries have <forward_list>, so we test for it instead.
    //   Thanks to Jonathan Wakely for devising the clever test for modern/ancient versions.
    // Visual Studio did not add template aliases until VS 2015 (v19.00). Compile failed under
    //   Visual Studio 2010 (v16.00) and Visual Studio 2012 (v17.00).
    #if (__cplusplus >= 201103L) || (_MSC_VER >= 1900)
    #  if defined(__clang__)
    #    if (__has_include(<forward_list>))
    #      define CRYPTOPP_HAVE_UNIQUE_PTR 1
    #    endif
    #  else
    #    define CRYPTOPP_HAVE_UNIQUE_PTR 1
    #  endif
    #endif

    // The result of below is a CryptoPP::auto_ptr in both cases
    #ifdef CRYPTOPP_HAVE_UNIQUE_PTR
      template<typename T>
       using auto_ptr = std::unique_ptr<T>;
    #else
      using std::auto_ptr;
    #endif

Then, in the source code:

    PK_Verifier::VerifyMessage(const byte*, size_t, const byte*, size_t) const
    {
        using CryptoPP::auto_ptr;
        auto_ptr<PK_MessageAccumulator> m(NewVerificationAccumulator());
        ...
    }

For testing on BSDs, Linux, OS X and Unix, you would build in one of four configurations to ensure correctness:

    # c++ -DNDEBUG -g2 -O3 ...
    export CXXFLAGS=""
    make

    # c++ -std=c++11 -DNDEBUG -g2 -O3 ...
    export CXXFLAGS="-std=c++11"
    make

    # c++ -stdlib=libc++ -DNDEBUG -g2 -O3 ...
    export CXXFLAGS="-stdlib=libc++"
    make

    # c++ -std=c++11 -stdlib=libc++ -DNDEBUG -g2 -O3 ...
    export CXXFLAGS="-std=c++11 -stdlib=libc++"
    make

-----

* https://github.com/weidai11/cryptopp/commit/52b7fbbca6274eddb9de6d7ee69ec058d34b3a76
* https://github.com/weidai11/cryptopp/commit/17fd66f2ff23253f9e713a1bff8fd2723ea5735e
* https://github.com/weidai11/cryptopp/commit/7ad12ab5b05eb2ee595e8900a5d2d101740c7a94
* https://github.com/weidai11/cryptopp/commit/4d8ac2db7c0d486a68f5b36658add615fb51e196
* https://github.com/weidai11/cryptopp/commit/8c630328656e1a7e5d8672d59b21c1dfa8ee6733
* https://github.com/weidai11/cryptopp/commit/60291122ae2fe6bd83a40a470e951f12f642fc73

**Comment by noloader** [comment]
> It seems challenging to simultaneously compile with no warnings both on C++11 platforms that deprecate `auto_ptr`, and C++03 platforms that don't have `unique_ptr`...

This observation was insightful. I found we were able to sidestep most of the warnings by switching to Crypto++'s `member_ptr`. It provides the same basic functionality we needed, and it has both `reset` and `release`. It does not require the [sometimes] convoluted logic to determine when C++11 is *really* available, and does not require template declaration support.

We have one place we could not cleanly cut over, and that was in `validat1.cpp` (around [line 575](http://www.cryptopp.com/docs/ref/validat1_8cpp_source.html#l00375)). So warnings went from "a lot" to "just one".
